src/common/options/rgw.yaml.in

```diff
@@ -2191,6 +2191,18 @@ options:
   services:
   - rgw
   with_legacy: true
+- name: rgw_bucket_stats_on_head_bucket
```

Comment:
i'm skeptical that this configurable is a good idea, given that clients now have to opt-in with request header x-rgw-get-bucket-stats

if the admin disables this, the only way for clients to get this information is by listing the entire bucket which is far more expensive

Reply:
right, the only scenario that bothers me is that HeadBucket is cheap to execute so it can be repeated many times in a second and that can be harmful to the system... even 1 per second for large buckets (with a high number of shards). but if the user goes through the listing bucket, by nature it takes time (because it needs to go one by one and follow the markers), so that can eliminate the case somehow (basically in a nutshell it won't be an easy task for the user so it will try alternative ways of e.g. the cloud-provider provided stats or stuff like that).

Reply:
i think i get what you're saying about the perceived cost of the different API calls. and while the cost of HeadBucket with stats is the same as a single paginated ListObjects request, rate limiting could effectively mitigate the users doing excessive bucket listings to calculate stats

but if a user has to add the x-rgw-get-bucket-stats request header when they want stats, that feels like an opportunity to change their perception about the cost. what if we call it x-rgw-get-expensive-bucket-stats? :) or document it somewhere like https://docs.ceph.com/en/latest/radosgw/s3/bucketops/ to say it has to sum over all index shards

if a service provider has their own preferred way to report stats, i guess the user will choose whichever method is better advertised to them. if HeadBucket really becomes a problem, a proxy could catch requests containing x-rgw-get-bucket-stats to reject them or bill them differently

Reply:
I see. Actually, I was looking at it as a DoS vulnerability as it can be executed by many at the same time without proper rate limiting. So, I thought having this control variable for the provider to control whether they want to provide such a feature or not might be easier than any custom resolution.
but I guess if we introduce the flag via query string rather than a HTTP header, it can be identified easier in the normal logs (without a need to dump header or special config) and then perhaps the provider can implement some control based on that arg.
what do you think?

Reply:
> Actually, I was looking at it as a DoS vulnerability as it can be executed by many at the same time without proper rate limiting.

by that logic, isn't ListObjects susceptible to the same attack? it can be expensive to produce sharded bucket listings, but i wouldn't want to call that an attack vector

that said, you've already implemented the config option for it. my only real objection is that it's disabled by default

Reply:
Right, that was a harsh statement. I agree :) My point was with listing, in one request, it doesn't open all shards and opens only one shard (one rados object—one OSD) at a time, so it kind of slows the user's unintended load.

> my only real objection is that it's disabled by default

Sure we can have it enabled by default. I wanted to make sure the admin knows what he has enabled, but I guess with the changelog we can inform this.

Reply:
> My point was with listing, in one request, it doesn't open all shards and opens only one shard (one rados object—one OSD) at a time, so it kind of slows the user's unintended load.

to produce an ordered list of ~1000 objects for each request, [RGWRados::cls_bucket_list_ordered()](https://github.com/ceph/ceph/blob/ce0d6fc/src/rgw/driver/rados/rgw_rados.cc#L9828) does have to visit all of the shards and sort the results

Reply:
ah right, thanks, I forgot that ordered listing is the default behavior for the API. OK, so this doesn't change anything :)
I'll be fine to drop the config if you think it doesn't worth it anymore.

Reply:
Alright, just a small nit, sorry for making it so long... What do you think of making this a query string rather than a header? This would be easier to identify in normal HTTP logs, I would say, than a header. Something like `?rgwx-get-bucket-stats`.

Reply:
sure. i don't think there's an easy way to get s3 clients/sdks to add a query param, so i wrote some glue in [service-2.sdk-extras.json](https://github.com/ceph/ceph/tree/3caa542/examples/rgw/boto3) to expose a `?read-stats` query param as `ReadStats`: https://github.com/ceph/ceph/pull/60847

Reply:
> sure. i don't think there's an easy way to get s3 clients/sdks to add a query param, so i wrote some glue in [service-2.sdk-extras.json](https://github.com/ceph/ceph/tree/3caa542/examples/rgw/boto3) to expose a `?read-stats` query param as `ReadStats`: #60847

@cbodley - Do we have a way to use this in s3-tests for `test_head_bucket_usage()`? I guess this will fail with this PR.

Reply:
@yuvalif just recently added s3-tests for https://github.com/ceph/ceph/pull/59808 that run with and without Ceph's service-2.sdk-extras.json extensions installed. the new qa/suites/rgw/bucket-logging suite is what runs with extensions

_ideally_ s3-tests would only cover the official apis so work against any s3-compatible backend, but that line has always been blurry. however, several of our extensions under https://github.com/ceph/ceph/tree/main/examples/rgw/boto3 also come with example python programs that use them. maybe we could find a way to run those examples as test cases in teuthology, instead of messing with s3-tests? the added benefit would be that we know those examples still work

Reply:
I thought `fails_on_aws` would do the trick for this but if we start to separate we might end up duplicating some bases but I don't think it should be something to fear.

But I just could come up with this as well to pass the arg: https://github.com/ceph/s3-tests/pull/605

Reply:
> But I just could come up with this as well to pass the arg: [ceph/s3-tests#605](https://github.com/ceph/s3-tests/pull/605)

i like that 👍